### PR TITLE
Dual topology hydration

### PR DESCRIPTION
**Pull request by proteneer**
This PR implements the relative hydration protocol using the RMSD based restraints. There's a fair bit of duplication in this code as we are trying to figure out the best way to refactor it to be a little cleaner. Some parts of note:

-setting up the core restraints is done in a pretty adhoc way right now
-examples/validate_relative_hydration.py is the main script that compares the ddG_ab against dG_a - dG_b on the hif2a set
-there's a duplicated estimator_abfe.py code that does the endpoint correction as part of the deltaG estimate. We should try and merge the two estimators into a single function later.
-model_rabfe is mostly a copy of model.py, but with some with auxiliary functionality like setting up restraints.

### CODE REVIEW

File: examples/validate_relative_hydration.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+    for epoch in range(100):
+
+        i = 0
+
+        mol_a = dataset.data[i]
+
+        dG_a = model_absolute.predict(ordered_params, mol_a, prefix='solvent_absolute_'+mol_a.GetProp('_Name'))
+
+    assert 0
```
Comment (badisa) [inline examples/validate_relative_hydration.py:158]: Should this be removed?

File: fe/model_utils.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                masses[j] -= multiplier*masses[i]
+                masses[i] += multiplier*masses[i]
+        else:
+            # do nothing
+            continue
+
+    return masses
+
+
+def generate_topology(objs, host_coords, out_filename):
```
Comment (badisa) [inline fe/model_utils.py:52]: nit: `generate_openmm_topology`?

File: fe/estimator_abfe.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        mean_du_dls.append(np.mean(result.du_dls))
+        all_grads.append(result.du_dps)
+
+    tibar_dG = 0
+    for lambda_idx in range(len(model.lambda_schedule) - 1):
+        fwd_work = ti_results[lambda_idx].du_dls
+        rev_work = -ti_results[lambda_idx+1].du_dls
+        delta_lamb = model.lambda_schedule[lambda_idx+1] - model.lambda_schedule[lambda_idx]
+        tibar, err = pymbar.BAR(model.beta*fwd_work, model.beta*rev_work)
+        tibar_dG += (tibar/model.beta)*delta_lamb
```
Comment (maxentile) [inline fe/estimator_abfe.py]: Documentation request: add comment at start of this block describing intent and any notable approximations, possibly with a reference. Alternately, move this free energy estimator into a separate function.

Some approximations I think should be noted:
* Estimate free energy difference between lam[0] and lam[-1] by summing up BAR estimates for subsequent pairs of states lam[i], lam[i+1]. Requires as input "delta U"s, i.e. `w_F=u(x, lam[i+1]) - u(x, lam[i])`.
* First-order Taylor approximation for `w_F=u(x, lam[i+1]) - u(x, lam[i])`, approximated as `w_F_prime=delta_lamb * du_dl(x, lam[i])`
* Interchange order of operations, replacing `BAR(w_F_prime, w_R_prime)` with `delta_lamb * BAR(w_F_prime/delta_lamb, w_R_prime/delta_lamb)`. (Probably doesn't make a huge difference, but should be noted that `b * BAR(w_F/b, w_R/b) != BAR(w_F, w_R)` in general.)

File: fe/model_utils.py (hunk 2 above)
Comment (proteneer) [reply fe/model_utils.py:52]: willfix

File: examples/validate_relative_hydration.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
+# this script validates the endpoint correction protocol on relative hydration free energies
+
+# This script repeatedly estimates the relative binding free energy of a single edge, along with the gradient of the
+# estimate with respect to force field parameters, and adjusts the force field parameters to improve tha accuracy
+# of the free energy prediction.
```
[the remark this answers is not in the document]
Comment (proteneer) [reply examples/validate_relative_hydration.py:5]: This PR actually doesn't train over the forcefield parameters. So I'm going to trim the text.

File: examples/validate_relative_hydration.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+array = Union[np.array, jnp.array]
+Handler = Union[AM1CCCHandler, LennardJonesHandler] # TODO: do these all inherit from a Handler class already?
+
+if __name__ == "__main__":
+
+    multiprocessing.set_start_method('spawn')
+
+    parser = argparse.ArgumentParser(
+        description="Absolute Binding Free Energy Testing",
```
[the remark this answers is not in the document]
Comment (proteneer) [reply examples/validate_relative_hydration.py:43]: willfix directly

File: examples/validate_relative_hydration.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    with open('ff/params/smirnoff_1_1_0_ccc.py') as f:
+        ff_handlers = deserialize_handlers(f.read())
+
+    forcefield = Forcefield(ff_handlers)
+    mols = [x for x in suppl]
+
+    dataset = Dataset(mols)
+
+    absolute_solvent_schedule = construct_absolute_lambda_schedule(cmd_args.num_windows)
+    relative_solvent_schedule = construct_absolute_lambda_schedule(cmd_args.num_windows-1)
```
[the remark this answers is not in the document]
Comment (proteneer) [reply examples/validate_relative_hydration.py:106]: Yes, but mainly for efficiency purposes, the internal free energy setup will add an extra stage at the end for the gas-phase simulation used in the endpoint correction. If we have 8 GPUs, and we have 8 windows, then we'd end up running 8+1 stages without the -1, so we'd have to do it in two passes.

File: fe/estimator_abfe.py (hunk 3 above)
Comment (proteneer) [reply fe/estimator_abfe.py]: Willfix

Comment (proteneer) [reply fe/estimator_abfe.py]: PS I haven't decided if I should include this yet, adding this here right now mainly for diagnostic purposes.

File: examples/validate_relative_hydration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxentile) [reply examples/validate_relative_hydration.py]: Resolved in https://github.com/proteneer/timemachine/pull/452/commits/5b9808fdba2c20c6d7130451040c744a033380ed

File: examples/validate_relative_hydration.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxentile) [reply examples/validate_relative_hydration.py:103]: Ahh, that makes sense -- thanks for explaining!

File: fe/estimator_abfe.py (hunk 3 above)
Comment (maxentile) [reply fe/estimator_abfe.py]: Resolved in https://github.com/proteneer/timemachine/pull/452/commits/a0dbff7c9b7327d12c2354397d9e5172cab61bb5 and https://github.com/proteneer/timemachine/pull/452/commits/212550143194be4583576cfce17a1af26474b5b4

**Review by badisa — APPROVED** [review]

**Review by maxentile — APPROVED** [review]

**Comment by proteneer** [comment]
Thanks for the comments, merging this now. I can finally combine this with binding code tomorrow.

**Pull request merged**